crates/red_knot/src/types.rs

```diff
@@ -142,6 +143,24 @@ impl TypeStore {
             intersection_id: id.intersection_id,
         }
     }
+
+    fn record_symbol_dependency(&self, from: (FileId, SymbolId), to: (FileId, SymbolId)) {
+        let (from_file_id, from_symbol_id) = from;
+        self.add_or_get_module(from_file_id)
+            .symbol_dependencies
+            .entry(from_symbol_id)
+            .or_default()
+            .insert(to);
+    }
+
+    fn record_module_dependency(&self, from: (FileId, SymbolId), to: ModuleName) {
```

Comment:
The case that you outline (same module name referring to a different module ID) is precisely why I think this needs to store a `ModuleName`. The dependency here is on "whatever this module name resolves to" -- if we have a dependency on `foo.bar`, and `foo/bar.py` doesn't change, but `foo/bar/__init__.py` is added so `foo.bar` now resolves there, we must invalidate the dependency on `foo.bar`. So the dependency is on the module name, not on the file ID.

Whether it could be on the module ID instead (i.e. type `Module`), or not, is unclear to me. That depends whether we guarantee stable 1:1 mapping between `FileId` and `Module`, or between `ModuleName` and `Module` (or neither?). If it's the latter, then I could store dependency to `Module` and it would be a little cheaper.

Reply:
The `ModuleId` mapping is stable between `ModuleName` and `ModuleId`. Although it is not guaranteed to be stable across runs (but that's not your concern, but something that the persistent caching layer must patch up).

Reply:
Hmm, maybe it would work out to keep this dependency to `FileId`, too? As long as in the case where a file is shadowed by a different file for the same module name, we always remove the now-shadowed file, invalidating all its data?

Reply:
We might. However, I'm also happy to use `ModuleId` when we know that it always comes from a `Module`. It gives the id more semantic meaning.